### DISCUSSION TITLE
Feature: Dynamic route parameters should not be set using definePageMeta

### DISCUSSION THREAD

**Discussion by adesombergh** [nuxt-modules/i18n#1730] [Ideas]
Hi, thanks for the great work with this module.

I stumbled into a problem with i18n module v8 and nuxt3 and I think it might be interesting to change this in the future, as it makes it difficult to use with dynamic content.

Please, correct me if i'm not doing in right.

### The Problem

According to the [Dynamic route parameters documentation](https://v8.i18n.nuxtjs.org/guide/lang-switcher#dynamic-route-parameters), to set translations for route params we should use `definePageMeta` like so:
```vue
<script setup>
definePageMeta({
  // ...
  nuxtI18n: {
    en: { id: 'my-post' },
    fr: { id: 'mon-article' }
  }
  // ...
})
</script>
```

However, the Nuxt 3 documentation about `definePageMeta` ([right here](https://nuxt.com/docs/guide/directory-structure/pages/#page-metadata:~:text=Much%20like%20defineEmits,reference%20imported%20bindings.)) says that **the page meta object cannot reference the component (or values defined on the component)**.

It means that the content inside `definePageMeta({})` is supposed to be **hard-coded and cannot be dynamic**, making it impossible to set the page route params dynamically (as they often are). The following code won't work:

```vue
<script setup>
// /pages/[slug].vue
const route = useRoute()
const { locale } = useI18n()
const { data } = await useFetch(
  `https://yourapi.com/${locale}/page/${route.params.slug}`
)

definePageMeta({ 
  nuxtI18n: data.value.other_languages // Throws an error because data is undefined
})
</script>
```

```json
{
  "other_languages": {
    "en": { "slug": "tech-specs"},
    "fr": { "slug": "specifications-techniques"},
  },
  ...more content ...
}
```

### A solution

Instead of using `definePageMeta`, maybe i18n module could give us a composable like `setParamsTranslations` or whatever,  that could be used like this:

```vue
<script setup>
// /pages/[slug].vue
const route = useRoute()
const { locale, setParamsTranslations } = useI18n()
const { data } = await useFetch(
  `https://yourapi.com/${locale}/page/${route.params.slug}`
)
setParamsTranslations(data.value.other_languages)
</script>
```

If anyone agrees I can create an issue (feature)

**Comment by madc** [comment]
We are just looking at the same issue. A composable would be very cool.

**Comment by adriaandotcom** [comment]
Trying this indeed does not work:

```js
const nuxtI18n = computed(() => articles?.value?.[0]?.languages || {});

definePageMeta({
  nuxtI18n,
});
```

**Reply by adriaandotcom** [reply]
If you aren't concerned about SSR, you can use this.

```vue
<script setup>
const route = useRoute()
definePageMeta({
  nuxtI18n: {}
})

const slug = ref()

// Let's say you have an article with the page slug
const page = computed(() => {
  if (!pages.value) return;
  slug.value = pages.value.slug;
  return pages.value;
});

watch(
  slug,
  () => { route.meta.nuxtI18n = nuxtI18n },
  { deep: true }
);
</script>
```

See https://github.com/nuxt-modules/i18n/discussions/1588#discussioncomment-4194858 for more info.

**Reply by adriaandotcom** [reply]
Even better:

https://github.com/nuxt-modules/i18n/issues/1736#issuecomment-1376617781

**Comment by BobbieGoede** [comment]
I have just released `rc.9` which includes the new [useSetI18nParams](https://i18n.nuxtjs.org/api/composables#useseti18nparams) composable, more detailed example of its usage [here](https://i18n.nuxtjs.org/guide/lang-switcher#dynamic-route-parameters).

The composable works in a similar way to your proposed solution, please let me know if it works for your or if you run into any issues!

**Reply by madc** [reply]
This made my day, thanks for working on this!

**Reply by adesombergh** [reply]
I have not tested it yet but it looks like it's exactly what we needed to set dynamically translated route params. Thank you for the good work!

**Comment by romainavalle** [comment]
Hey guys

I think I have found a workaround for SSR dynamic slugs.
I would love feedbacks cause it is a bit hacky.
I delay the rendering of the lang-switcher by either waiting a callback event to know that the I18nParams hd been changed or just a timeout.

https://gist.github.com/romainavalle/638dcffb657fe49857a00a45aa3ecf48
